### ISSUE TITLE
Attention Computation Question

### ISSUE THREAD

**Issue by ChrisSsak** [AmeenAli/HiddenMambaAttn#7]
Hello,

Thank you for the amazing work!

I had a question regarding attenion in [compute_middle_attention_vector](https://github.com/AmeenAli/HiddenMambaAttn/blob/4963a481fe2a51b7cf2a45199d017aa71fc3ddc6/mamba-1p1p1/mamba_ssm/ops/selective_scan_interface.py#L719), if I do not add a cls token to my sequence, either at frond, middle or end. Does torch.sum(C * dA * dB, axis=-1), also count as self attention over the entire sequence? or is a cls token needed and the formula in the paper and the repository must be followed exactly as stated?

Thank you!

**Comment by AndssY** [comment]
Have you figured out this problem? I am also troubled by this issue. 
I didn't use cls token either, and I also want to get 'self attention' for the entire sequence.

**Comment by Itamarzimm** [comment]
Hi @ChrisSsak @AndssY, thanks for your questions! 
The CLS token is not necessary for our formulation, and our view of Mamba as self-attention over the entire sequence does not rely on any special conditions or assumptions. [The last commit ](https://github.com/AmeenAli/HiddenMambaAttn/commit/929d403ed3cc6eb28c8db1de8ed9208fbad43d81) includes computations of the entire (hidden) attention matrices for each channel and layer using the [compute_attn_matrix_fn](https://github.com/AmeenAli/HiddenMambaAttn/blob/929d403ed3cc6eb28c8db1de8ed9208fbad43d81/mamba-1p1p1/mamba_ssm/ops/selective_scan_interface.py#L741C5-L741C28) function. We have also added visualizations of these matrices in the [notebook](https://github.com/AmeenAli/HiddenMambaAttn/blob/929d403ed3cc6eb28c8db1de8ed9208fbad43d81/vim/vmamba_xai.ipynb).

**Comment by ChrisSsak** [comment]
Thank you so much!

I just had one two brief questions as I am really confused.

1- If I want to compute the attention as a vector using compute_middle_attn_vector over the entire sequence, I am quite confused about the cls_pos = 98, should I set cls_pos = L instead?

2- From the paper, Q K H are equivalent to QKV, theoretically, thus, theoretically, If I use C from one input and delta, delta_bias, A and B from another input, wouldn't that be equivalent to cross attention instead of self attention?


Apologies for the long questions.

Thank you!

**Comment by Itamarzimm** [comment]
Hi @ChrisSsak ,

1. **CLS Positioning:** Yes, in your case, you should set `cls_pos = L`. It's important to note that in Vision Mamba, the cls token is centrally located. To effectively provide interpretability tools for this configuration, such settings are essential.

2. **Cross-Hidden-Attention:** The concept of cross-hidden-attention is exciting! We have explored similar ideas in our lab and have observed promising initial results. However, relatively similar mechanism was introduced in the [Sigma paper](https://arxiv.org/pdf/2404.04256) (see equations 9-11) before we finished the multimodal project (although it did not use hidden attention as an underlying motivation).

Thanks for your questions!
